### PR TITLE
Split core module to core, aws-auth and aws-core

### PR DESCRIPTION
**Pull request by zoewangg**
## Description
Splitting core module to auth, core, aws-core and profiles modules.
- regions: regions classes
- profiles: contains profile classes.
- auth: contains credentials, aws signer classes
- aws-core: core runtime classes used by aws java clients.
- core: core runtime classes.

**Dependencies**:
profiles -> core
regions -> profiles
auth -> regions
aws-core -> auth
services -> aws-core
Apigatway generated clients -> core
Apigatway generated clients using IAM role -> core, auth

## Motivation and Context
Related to #27 
Note: This is the first task to move out AWS related classes.
The rest of the tasks left are:

- ResponseHandler and Json error response handler classes are still dealing with aws headers and need to be moved to aws-core.
- `CRC32_FROM_COMPRESSED_DATA_ENABLED` needs to be moved to aws-core module once we replace the legacy`HttpResponse` with `SdkHttpFullResponse`

## Testing
`mvn clean install` and all integration tests were passed

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

**Comment by zoewangg** [comment]
This motivation of splitting core module is to break out AWS related concepts(AWS credentials, AWS signers, etc) to a separate module so that the clients generated from API gateway API just needs to depend on core module rather than pulling the whole AWS-specific logic.

As to the naming, using `aws-auth` makes is clear that it contains AWS related authentication classes and applies to all AWS services. The clients generated from API gateway don't need to depend on it unless they use IAM roles for authentication.

I thought about updating `core` to `sdk-core` to differentiate from `aws-core`, but don't have strong opinions and open to discussion.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/AwsExecutionAttributes.java [hunk 1]
```diff
@@ -32,39 +34,14 @@
                      + "4. The remaining public AWS stuff should be moved to the AWS module. "
                      + "We should also consider making some of the SDK/AWS-owned set of attributes part of the immutable context "
                      + "if we don't want the interceptors to modify them.")
-public final class AwsExecutionAttributes {
+public interface AwsExecutionAttributes extends SdkExecutionAttributes {
```
Comment (shorea) [inline aws-auth/src/main/java/software/amazon/awssdk/awsauth/AwsExecutionAttributes.java]: This seems to be more of an awscore thing rather than auth specific.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/AwsSystemSetting.java [hunk 2]
```diff
@@ -43,9 +35,8 @@
 public enum AwsSystemSetting implements SystemSetting {
```
Comment (shorea) [inline aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/AwsSystemSetting.java]: Belongs in awscore

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/HttpCredentialsProvider.java [hunk 3]
```diff
@@ -132,7 +132,7 @@ public void close() {
         private String asyncThreadName;
 
         /**
-         * Created using {@link #builder()}.
+         * Created using {@link #build()}.
```
Comment (shorea) [inline aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/HttpCredentialsProvider.java]: Nitpicky but this seemed more correct before. Comment is mostly useless though anyways.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/internal/ConnectionUtils.java [hunk 4]
```diff
@@ -13,7 +13,7 @@
  * permissions and limitations under the License.
  */
 
-package software.amazon.awssdk.core.internal.net;
+package software.amazon.awssdk.awsauth.credentials.internal;
```
Comment (shorea) [inline aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/internal/ConnectionUtils.java]: For ConnectionUtils we should consider refactoring to use the SDK HTTP client instead of URL connection. Also we should consider putting the ec2 metadata "client" into it's own module.

**Comment by kiiadi** [comment]
I suppose that makes the thought process a little clearer - but from a customer perspective do you think they're going to notice the subtly between `software.amazon.awssdk:auth` and `software.amazon.awssdk:aws-auth`? As I say, we're already in the `awssdk` groupId.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/profile/Profile.java [hunk 5]
```diff
@@ -13,7 +13,7 @@
  * permissions and limitations under the License.
  */
 
-package software.amazon.awssdk.auth.profile;
+package software.amazon.awssdk.awsauth.credentials.profile;
```
Comment (shorea) [inline aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/profile/Profile.java]: The profile stuff may be better place in it's own module. It isn't just credentials and region and we may expand on it in the future to include other profile configuration.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/regions/Region.java [hunk 6]
```diff
@@ -13,7 +13,7 @@
  * permissions and limitations under the License.
  */
 
-package software.amazon.awssdk.core.regions;
+package software.amazon.awssdk.awsauth.regions;
```
Comment (shorea) [inline aws-auth/src/main/java/software/amazon/awssdk/awsauth/regions/Region.java]: Auth and region are related but not necesarrily coupled. Should region related stuff be in a separate module?

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/AwsExecutionAttributes.java (hunk 1 above)
Comment (zoewangg) [reply aws-auth/src/main/java/software/amazon/awssdk/awsauth/AwsExecutionAttributes.java]: This is used by Aws4Signer for now to get the credentials from ExecutionAttributes. I guess this will be moved around once we start working on signer refactor.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/AwsSystemSetting.java (hunk 2 above)
Comment (zoewangg) [reply aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/AwsSystemSetting.java]: This is used by some credential providers, eg:`SystemSettingsCredentialsProvider` is using it to locate the credentials.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/regions/Region.java (hunk 6 above)
Comment (zoewangg) [reply aws-auth/src/main/java/software/amazon/awssdk/awsauth/regions/Region.java]: Yes, we can further split aws-auth module to `region`, `profile`, `credentials`, `signer`, I'll try and see how it looks like, not sure if they have circular dependency right now.

File: core/src/main/java/software/amazon/awssdk/core/SdkRequestOverrideConfig.java [hunk 7]
```diff
@@ -15,314 +15,44 @@
...
-
-    /**
-     * Optional additional headers to be added to the HTTP request.
-     *
-     * @return The optional additional headers.
-     */
-    public Optional<Map<String, List<String>>> headers() {
-        return Optional.ofNullable(headers);
-    }
+public final class SdkRequestOverrideConfig extends RequestOverrideConfig {
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/core/SdkRequestOverrideConfig.java]: Why is this needed?

File: core/src/main/java/software/amazon/awssdk/core/client/SdkAsyncClientHandlerImpl.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Default implementation of {@link AsyncClientHandler}.
+ */
+@Immutable
+@ThreadSafe
+@SdkProtectedApi
+final class SdkAsyncClientHandlerImpl extends BaseAsyncClientHandler implements AsyncClientHandler {
+
+    @ReviewBeforeRelease("Should this be migrated to use a params object, particularly because it crosses module boundaries?" +
+                         "We might also need to think about how it will work after AWS/SDK are split.")
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/core/client/SdkAsyncClientHandlerImpl.java]: Is this comment now addressed?

File: core/src/main/java/software/amazon/awssdk/core/client/builder/ClientAsyncHttpConfiguration.java [hunk 9]
```diff
@@ -68,7 +68,7 @@ private ClientAsyncHttpConfiguration(DefaultHttpConfigurationBuilder builder) {
      * to ease resolution of the client. Returns an empty {@link Optional} if neither is set.
      */
     @SdkInternalApi
-    Optional<Either<SdkAsyncHttpClient, SdkAsyncHttpClientFactory>> toEither() {
+    public Optional<Either<SdkAsyncHttpClient, SdkAsyncHttpClientFactory>> toEither() {
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/core/client/builder/ClientAsyncHttpConfiguration.java]: Hmm I don't like this. Is there anything we can do about it? Maybe a ReviewBeforeRelease annotation as I believe we want to flatten this out anyways which could make this cleaner.

File: core/src/main/java/software/amazon/awssdk/core/http/AmazonSyncHttpClient.java [hunk 10]
```diff
@@ -55,15 +55,15 @@
 
 @ThreadSafe
 @SdkInternalApi
-public class AmazonHttpClient implements SdkAutoCloseable {
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/core/http/AmazonSyncHttpClient.java]: Can you add a ReviewBeforeRelease annotation to come up with a better name for this and the async variant?

File: core/src/main/java/software/amazon/awssdk/core/http/pipeline/stages/SigningStage.java [hunk 11]
```diff
@@ -81,21 +80,18 @@ private Signer newSigner(final SdkHttpFullRequest request, RequestExecutionConte
...
-     * {@link
-     * CanHandleNullCredentials}).
+     * We sign if a signer is provided is not null.
      *
      * @return True if request should be signed, false if not.
      */
-    private boolean shouldSign(Signer signer, AwsCredentials credentials) {
-        return signer != null && (credentials != null || signer instanceof CanHandleNullCredentials);
+    private boolean shouldSign(Signer signer) {
+        return signer != null;
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/core/http/pipeline/stages/SigningStage.java]: This seems like a behavior change. Is this okay?

File: core/src/main/java/software/amazon/awssdk/core/interceptor/SdkExecutionAttributes.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+ * Contains attributes attached to the execution. This information is available to {@link ExecutionInterceptor}s and
+ * {@link Signer}s.
+ */
+public interface SdkExecutionAttributes {
+
+    /**
+     * The key under which the request config is stored.
+     */
+    @ReviewBeforeRelease("RequestConfig feels pretty internal. Can we just expose parts of it?")
+    ExecutionAttribute<RequestOverrideConfig> REQUEST_CONFIG = new ExecutionAttribute<>("RequestConfig");
```
Comment (shorea) [inline core/src/main/java/software/amazon/awssdk/core/interceptor/SdkExecutionAttributes.java]: This comment doesn't seem to apply anymore since RequestOverrideConfig is a public class.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/AwsExecutionAttributes.java (hunk 1 above)
Comment (shorea) [reply aws-auth/src/main/java/software/amazon/awssdk/awsauth/AwsExecutionAttributes.java]: Shirley the awsauth module should depend on awscore?

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/HttpCredentialsProvider.java (hunk 3 above)
Comment (zoewangg) [reply aws-auth/src/main/java/software/amazon/awssdk/awsauth/credentials/HttpCredentialsProvider.java]: yup, removed

**Comment by zoewangg** [comment]
Yes, this PR is just the first step and there are still some remaining AWS functionality such as parsing aws headers in response handler, stax response marshaller, retry related stuff as you mentioned. To keep PR smaller(it's already huge as it is..), I will create separate PR for each of the remaining task.

File: aws-auth/src/main/java/software/amazon/awssdk/awsauth/regions/Region.java (hunk 6 above)
Comment (zoewangg) [reply aws-auth/src/main/java/software/amazon/awssdk/awsauth/regions/Region.java]: profile classes are moved to profiles module via commit https://github.com/aws/aws-sdk-java-v2/pull/475/commits/70825f422e21e65393d91394a052f67daa3e0105. 

As to regions and credential classes, they are tightly coupled now because they share common functionalities to to retrieve regions/creds from local EC2 instances in `HttpCredentialUtils`, `endpointRetry` and other classes, unless we break out those class to a separate module, it's difficult to separate them. If we are going with that route, we might end up having plenty "micro" modules and it could increase our maintenance burden and unlike profiles, regions and creds seem to be always used together.
